linkerd/opencensus/src/lib.rs

```diff
@@ -178,30 +179,47 @@ where
                     let req = grpc::Request::new(request_rx);
                     trace!("Establishing new TraceService::export request");
                     this.metrics.start_stream();
-                    let _rsp = Box::pin(async move { svc.export(req).await });
+                    let rsp = Box::pin(async move { svc.export(req).await });
                     State::Sending {
                         sender: request_tx,
                         node: Some(this.node.clone()),
-                        _rsp,
+                        rsp: Some(rsp),
                         metrics: this.metrics.clone(),
                     }
                 }
                 State::Sending {
                     ref mut sender,
                     ref mut node,
                     ref mut metrics,
+                    ref mut rsp,
                     ..
                 } => {
-                    match ready!(Self::poll_send_spans(
-                        this.spans,
-                        sender,
-                        node,
-                        *this.max_batch_size,
-                        metrics,
-                        cx,
-                    )) {
-                        Ok(()) => return Poll::Ready(()),
-                        Err(()) => State::Idle,
+                    let mut idle = false;
+                    if let Some(mut f) = rsp.take() {
+                        match f.poll_unpin(cx) {
+                            Poll::Ready(Ok(_)) => {}
+                            Poll::Pending => *rsp = Some(f),
+                            Poll::Ready(Err(error)) => {
+                                tracing::debug!(%error, "response future failed, sending a new request");
+                                idle = true;
+                            }
+                        }
+                    }
```

Comment:
I would nitpick this all more re: readability, but I'd rather just reewrite it with async...

Reply:
yup, i didn't spend a whole lot of time making it nice for that reason --- currently working on the much cleaner rewrite